### PR TITLE
Improve the documentation of our conventions

### PR DESCRIPTION
**Pull request by kyouko-taiga**
Addresses #949 and #968.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -28,17 +28,15 @@ If it hasn't, you can create a new issue with the label `bug` with the following
 - a description of the bug that you encountered;
 - the steps to reproduce it (typically in the form of a program);
 - the expected result;
-- the actual result; and
-- the version of your operating system and the version of the Hylo compiler you are using.
+- the actual result;
+- the version of your operating system; and
+- the commit hash of the code you are using.
```
Comment (dabrahams) [inline CONTRIBUTING.md:33]:
```suggestion
- the commit hash and public repository of the code you are using.
```

File: CONVENTIONS.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+It is intended for developers who wish to contribute to the Hylo compiler or simply read its sources.
+
+## Code formatting and conventions
+
+The Swift source code in this repository follows the [Swift API Design guidelines](https://www.swift.org/documentation/api-design-guidelines/).
+In addition, we maintain a set of code formatting conventions checked with `swift-format`.
+There is a CI job that checks the formatting; this job will not auto-correct the formatting, it will just ensure that the formatting is correct.
+
+Contributors are asked to format all the code they submit to the repository.
+We use `swift-format` tagged with the commit [0bc2f03](https://github.com/apple/swift-format/commit/0bc2f0381c72d66a949254af22208a81377cf717).
```
Comment (dabrahams) [inline CONVENTIONS.md:13]: I thought we were finally not tied to a particular commit of swift-format? @lucteo ?

Comment (dabrahams) [reply CONVENTIONS.md:13]: These instructions are not primarily about policy but setting up your development environment. IMO keeping that part in a separate document is appropriate.

File: CONVENTIONS.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+In general, a file is named after the main entity that it defines.
+A file that extends an existing type with a protocol (Swift) or trait (Hylo) conformance is named with a combination of the type name and the protocol or trait name, joined with a plus (`+`) sign.
+For more complex situations, exercise your best judgment.
+
+For example:
+- A Swift file defining a type named `MyType` is named `MyType.swift`.
+- A Swift file defining how `MyType` conforms to `Equatable` is named `MyType+Equatable.swift`.
+
+Avoid defining multiple types, protocols, or traits in the same file unless they are scoped by a main entity or meant to be used only in that file.
```
Comment (dabrahams) [inline CONVENTIONS.md:70]:
```suggestion
Avoid defining multiple types, protocols, or traits in the same file unless they are scoped by a main entity or meant to be used only in that file.
Usually, conformances are small and are defined in the same file as the type to which they apply.
```

File: CONVENTIONS.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+### Traps
+
+#### Unimplemented features
+
+It may be sometimes beneficial to stub the API of a feature before it is implemented.
+In those cases, use the function `UNIMPLEMENTED` from the `Utils` module to mark a particular code path as being unimplemented.
```
Comment (dabrahams) [inline CONVENTIONS.md:78]:
```suggestion
In those cases,  file an issue for the missing feature (include a reproducer) and use `UNIMPLEMENTED("description of feature #issue-number")` from the `Utils` module to mark the unimplemented code path.
Many uses of `UNIMPLEMENTED()` predate this policy and don't include the string argument; if you hit one, please file the issue and submit a PR that adds the appropriate string.
```

File: CONVENTIONS.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+#### Unreachable code paths
+
+Use the function `unreachable` to document that a particular code path is assumed to be unreachable because of some code logic that the type checker cannot verify.
```
Comment (dabrahams) [inline CONVENTIONS.md:82]:
```suggestion
Call `unreachable()` (from the `Utils` module) on any code path that indicates a bug in the Hylo compiler if ever reached. 
```

File: CONVENTIONS.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+## Reporting bugs
+
+Please attach a commit hash to issues reporting a bug.
```
Comment (dabrahams) [inline CONVENTIONS.md:86]:
```suggestion
Please attach a commit hash to issues reporting a bug.
Please try to include a minimal Hylo reproducer.
```

**Review by dabrahams — APPROVED** [review]
Thanks! IMO tooling setup belongs elsewhere.  Please see suggestions.

File: CONVENTIONS.md (hunk 5 above)
Comment (kyouko-taiga) [reply CONVENTIONS.md:82]: Qualification is needless in 99% of the cases. I'm not sure we should recommend `Utils.unreachable()`.

Comment (dabrahams) [reply CONVENTIONS.md:83]: I thought that might be an issue.  I wasn't intending to recommend the use of qualification.
```suggestion
Call `unreachable()` (defined in `Utils`) on any code path that indicates a bug in the Hylo compiler if ever reached. 
```

File: CONVENTIONS.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+Avoid defining multiple types, protocols, or traits in the same file unless they are scoped by a main entity or meant to be used only in that file.
+Usually, conformances are small and are defined in the same file as the type to which they apply.
+
+### Traps
+
+#### Unimplemented features
+
+It may be sometimes beneficial to stub the API of a feature before it is implemented.
+In those cases,  file an issue with a reproducer for the missing feature and call `Utils.UNIMPLEMENTED("description of feature #issue-number")` to mark the unimplemented code path.
```
Comment (dabrahams) [inline CONVENTIONS.md:78]:
```suggestion
In those cases,  file an issue with a reproducer for the missing feature and call `UNIMPLEMENTED("description of feature #issue-number")` (defined in `Utils`) to mark the unimplemented code path.
```

**Pull request merged**